### PR TITLE
[12634] Sign in Service Private Key JWT flow

### PR DESCRIPTION
**Pull request by bosawt**
## Summary

- This PR adds the Private Key JWT flow to the sign in service. This allows a backend client to facilitate authentication with sign in service, and is an additional configuration option to the current PKCE flow.

To use this flow, first set up a new `SignIn::ClientConfig` with:
`pkce = false`
` certificates = [<certificate>]`

- This is an example of how you can create a new client config in a rails console:

```
 client_config = SignIn::ClientConfig.find_or_initialize_by(client_id: 'your-client-id')
client_config.update!(authentication: SignIn::Constants::Auth::COOKIE,
             anti_csrf: false,
             redirect_uri: 'http://localhost:3001/auth/login/callback',
             access_token_duration: SignIn::Constants::AccessToken::VALIDITY_LENGTH_SHORT_MINUTES,
              access_token_audience: 'va.gov',
              pkce: false,
              certificates: ["your-public-cert"],
              logout_redirect_uri: 'http://localhost:3001',
              refresh_token_duration: SignIn::Constants::RefreshToken::VALIDITY_LENGTH_SHORT_MINUTES)
```

- The certificate described above is created with this command: `openssl req -nodes -x509 -days 365 -newkey rsa:2048 -keyout private.pem -out public.crt`, copy the `public.crt` file as a string, into an array entry in the `certificates` column on the client config

- Next, authenticate with this new client config (in your browser)
`localhost:3000/v0/sign_in/authorize?type=logingov&acr=min&client_id=my_new_private_key_jwt_client`
- Take note of the `code` param in the response
- Create a JWT to send for a `/token` request, using the `private.pem` file as a `private_key` :
`JWT.encode({iss: 'my_new_private_key_jwt_client', aud: 'http://127.0.0.1:3000/v0/sign_in/token', sub: 'my_new_private_key_jwt_client', jti: SecureRandom.hex, exp: 1.month.since.to_i}, OpenSSL::PKey::RSA.new(File.read(private_key)), 'RS256')`
- Send a request to the sign in service `/token` route, using 'code' from above, and the 'encoded jwt' from above:
`curl -X POST localhost:3000/v0/sign_in/token -H 'Content-Type: application/json' -d '{"grant_type": "authorization_code", "code": "<some-code>", "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer", "client_assertion": "<some-client-assertion-jwt>"}'`
- Confirm authentication was successful

## Related issue(s)
- department-of-veterans-affairs/vets-api#12634

## Testing done

- Created a certificate and a new client for sign in service, then confirmed authentication with private key jwt flow
- Confirmed when client is set to PKCE, private key jwt flow cannot be used
- Confirmed when client is set to PKCE=false, then PKCE flow cannot be used

## What areas of the site does it impact?
Authentication

## Acceptance criteria

- [ ] Create your own client config with pkce=true, expectation is that this flow should work exactly as the existing Sign in Service flow
- [ ] Create a client config with pkce=false, additionally, add a certificate to the certificates column, using the step outlined above
- [ ] Attempt authentication with client that is configured with pkce=false, using the private_key_jwt flow
- [ ] Confirm any single client config can only be used with either pkce, or private_key_jwt, not both

### CODE REVIEW

File: app/services/sign_in/client_assertion_validator.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    def validate_audience
+      unless decoded_client_assertion.aud.match(token_route)
+        raise Errors::ClientAssertionAttributesError.new message: 'Client assertion audience is not valid'
+      end
+    end
+
+    def token_route
+      "#{Settings.hostname}#{Constants::Auth::TOKEN_ROUTE_PATH}"
```
Comment (rileyanderson) [inline app/services/sign_in/client_assertion_validator.rb:47]: This is missing a slash - 
```suggestion
      "#{Settings.hostname}/#{Constants::Auth::TOKEN_ROUTE_PATH}"
```

File: app/services/sign_in/constants/auth.rb [hunk 2]
```diff
@@ -25,12 +25,15 @@ module Auth
...
       GRANT_TYPE = 'authorization_code'
+      CLIENT_ASSERTION_TYPE = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'
+      CLIENT_ASSERTION_ENCODE_ALGORITHM = 'RS256'
       IAL = [IAL_ONE = 1, IAL_TWO = 2].freeze
       INFO_COOKIE_NAME = 'vagov_info_token'
       JWT_ENCODE_ALGORITHM = 'RS256'
       LOA = [LOA_ONE = 1, LOA_THREE = 3].freeze
       REFRESH_ROUTE_PATH = '/v0/sign_in/refresh'
       REFRESH_TOKEN_COOKIE_NAME = 'vagov_refresh_token'
+      TOKEN_ROUTE_PATH = 'v0/sign_in/token'
```
Comment (rileyanderson) [inline app/services/sign_in/constants/auth.rb:36]: Or you could put the slash here 🤷‍♂️ 
```suggestion
      TOKEN_ROUTE_PATH = '/v0/sign_in/token'
```

**Comment by bramleyjl** [comment]
I was able to follow the above instructions and successfully make it through a token response:
```
2023-05-18 15:12:15.990691 D [77220:puma srv tp 004] ActiveRecord::Base --   SignIn::ClientConfig Load (0.1ms)  SELECT "client_configs".* FROM "client_configs" WHERE "client_configs"."client_id" = $1 LIMIT $2  [["client_id", "jwt_auth_test"], ["LIMIT", 1]]
2023-05-18 15:12:15.990931 D [77220:puma srv tp 004] ActiveRecord::Base --   ↳ app/models/sign_in/access_token.rb:92:in `client_config'
2023-05-18 15:12:15.992972 D [77220:puma srv tp 004] Rails --   Encrypt Data Key (0.0ms)  Context: 
2023-05-18 15:12:15.995695 I [77220:puma srv tp 004] Rails -- [SignInService] [V0::SignInController] token -- { :user_uuid => "e444837a-e88b-4f59-87da-10d3c74c787b", :session_id => "a7f31c0a-c35a-4b77-b5f2-72087db7aba1", :token_uuid => "41faaa46-5f51-40b2-8747-659faf423c4c" }
2023-05-18 15:12:15.997630 I [77220:puma srv tp 004] Rails -- Rendered ActiveModel::Serializer::Null with Hash (0.03ms)
2023-05-18 15:12:15.997762 I [77220:puma srv tp 004] Rails -- Completed 200 OK in 148ms (Views: 1.6ms | ActiveRecord: 5.7ms | Allocations: 99747)
```

The `ClientConfig` creation object was very helpful in testing this and should probably be added to the documentation:
```
vaweb = SignIn::ClientConfig.find_or_initialize_by(client_id: '<client_id here>')
vaweb.update!(authentication: SignIn::Constants::Auth::COOKIE,
              anti_csrf: false,
              redirect_uri: 'http://localhost:3001/auth/login/callback',
              access_token_duration: SignIn::Constants::AccessToken::VALIDITY_LENGTH_SHORT_MINUTES,
              access_token_audience: 'va.gov',
              pkce: false,
              certificates: ["<public certificate here>"],
              logout_redirect_uri: 'http://localhost:3001',
              refresh_token_duration: SignIn::Constants::RefreshToken::VALIDITY_LENGTH_SHORT_MINUTES)
```

**Comment by bosawt** [comment]
> vaweb = SignIn::ClientConfig.find_or_initialize_by(client_id: '<client_id here>')
> vaweb.update!(authentication: SignIn::Constants::Auth::COOKIE,
>               anti_csrf: false,
>               redirect_uri: 'http://localhost:3001/auth/login/callback',
>               access_token_duration: SignIn::Constants::AccessToken::VALIDITY_LENGTH_SHORT_MINUTES,
>               access_token_audience: 'va.gov',
>               pkce: false,
>               certificates: ["<public certificate here>"],
>               logout_redirect_uri: 'http://localhost:3001',
>               refresh_token_duration: SignIn::Constants::RefreshToken::VALIDITY_LENGTH_SHORT_MINUTES)

I'll move this into the PR description, it's a good callout

**Review by rileyanderson — APPROVED** [review]
Looks good! I was able to auth with both `id.me` and `login.gov` following your instructions and with the [sample app ](https://github.com/department-of-veterans-affairs/sign-in-service-client-sinatra/pull/24). 🚀

**Comment by tynyttie** [comment]
It appears satisfactory. I successfully completed authentication using both id.me and login.gov. I suggest including a reminder in the instructions to complete the prompts for openssl. Failure to answer the prompt questions will result in the non-creation of the .crt file.

**Review by tynyttie — APPROVED** [review]
It looks good. I successfully completed authentication using both id.me and login.gov. I suggest including a reminder in the instructions to complete the prompts for openssl. Failure to answer the prompt questions will result in the non-creation of the .crt file.

**Comment by joeniquette** [comment]
@tynyttie Can you share your config file and which test accounts you used with @achievedaniel so he can test? Please dont share them in this github ticket, I dont want the credentials to go to the public side. Share via slack please.

**Comment by tynyttie** [comment]
@joeniquette Will do.

**Comment by joeniquette** [comment]
Confirmed was able to login with login.gov account 
Screenshot 2023-05-23 at 14 02 14

**Review by joeniquette — APPROVED** [review]

**Pull request merged**